Add decentralized live migration documentation (#914)
kubevirt/user-guide: docs/compute/decentralized_live_migration.md

### Decentralized live migration Feature Gate

Decentralized live migration support must be enabled in the feature gates to be available. The [feature gates](../cluster_admin/activating_feature_gates.md#how-to-activate-a-feature-gate)
field in the KubeVirt CR must be expanded by adding the `DecentralizedLiveMigration` to it.

# Architecture

**Compute live migrations**

Compute live migrations are characterized by a single `VirtualMachineInstance` resource and a single `VirtualMachineInstanceMigrations` resource. A migration is initiated by creating a `VirtualMachineInstanceMigration` resource, with coordination occurring via the VirtualMachineInstance's status, as shown in Figure 1.

Figure 1:
![Figure 1](live-migration.png)

**Storage live migrations**

Storage live migrations, a variation of compute live migrations, involve copying the virtual machine's state and disk contents over the migration network. More information on initiating a storage live migration can be found in this [document](https://kubevirt.io/user-guide/storage/volume_migration/).

**Decentralized live migrations**

> Decentralized live migrations are a variation of storage live migrations. They involve two `VirtualMachineInstances` and two `VirtualMachineInstanceMigrations`. Similar to storage live migrations, disk contents are copied over the network to the receiving virtual machine. However, the key difference is that the receiving virtual machine has a completely separate `VirtualMachineInstance`. This necessitates synchronization between the source and destination to coordinate the migration, as illustrated in Figure 2.

Reviewer: OK, the "subject" was not immediately clear - if we are speaking about VMI statuses, then I get it. 

Consider re-wording the sentence.

Figure 2:
![Figure 2](decentralized-live-migration.png)

A dedicated synchronization controller facilitates communication between the source and target `VirtualMachineInstances`, as depicted in Figure 2. This design enables the reuse of most existing live migration mechanisms developed for `compute` migrations.

# Required resources

> To migrate a virtual machine (VM) between namespaces or clusters, you must create both a target `VirtualMachine` resource and a target `VirtualMachineInstanceMigration` resource. A decentralized live migration cannot be achieved using only a `VirtualMachineInstance` resource.

Reviewer 2: "To migrate a virtual machine (VM)"
I appreciate this, however it would be better off at the top for first use. 
Funnily enough, you've used "virtual machine" for nearly all instances throughout this file. I'm largely ambivalent, since there's a case for both (and 'VM' outweighs 'virtual machine' quite heavily across the docs as a whole) but we may as well be consistent in each markdown file. For instance, L6 sneaks in a 'VM' but then reverts back to 'virtual machine'. 
I'd call this a nit :)

> The target cluster/namespace must possess all necessary resources for the VM to operate. Some resources may not be an exact match on the target. For example, the storage class available on the target might differ from the source. Since decentralized live migration is a type of storage live migration, it is possible to change the storage class on the target and successfully complete the migration. The same applies to networking, though connectivity may be disrupted depending on the network configuration.

Author: Technically that layer is not required, the user can create all the resources themselves. I guess the user would be the orchestration layer at that point.

Additionally, auxiliary resources, such as a `Secret` containing an SSH key or a `ConfigMap` with configuration data, must also be copied and made available on the target cluster.

## Target virtual machine

> The target virtual machine is a copy of the source virtual machine, with its disks and networking configured to match the target cluster/namespace. The `runStrategy` must be set to `WaitAsReceiver`, signaling to KubeVirt that this is a special target virtual machine and that standard virtual machine commands (start/stop) will not function until the migration is completed. Consider the following source virtual amchine running on cluster 1.

Reviewer: caught a typo:

```suggestion
The target virtual machine is a copy of the source virtual machine, with its disks and networking configured to match the target cluster/namespace. The `runStrategy` must be set to `WaitAsReceiver`, signaling to KubeVirt that this is a special target virtual machine and that standard virtual machine commands (start/stop) will not function until the migration is completed. Consider the following source virtual machine running on cluster 1.
```

Author: Thanks

	blank: {}
  storage:
    storageClassName: <other storage class>
    resources:
      requests:
        storage: 10Gi
```

Target Virtual Machine:

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  annotations:
	kubevirt.io/restore-run-strategy: Always
  name: example-decentralized-lm-vm
spec:
  runStrategy: WaitAsReceiver
  template:
	metadata:
  	  creationTimestamp: null
	spec:
  	  architecture: amd64
  	  domain:
    	    devices:
    	    autoattachGraphicsDevice: false
    	  disks:
    	  - disk:
             bus: virtio
      	    name: disk0
    	  interfaces:
    	  - masquerade: {}
      	    name: default
    	  machine:
      	    type: q35
    	  resources:
      	    requests:
             memory: 128Mi
	networks:
	- name: default
  	  pod: {}
	volumes:
  	  - dataVolume:
      	    name: example-decentralized-lm-vm-dv
    	  name: disk0
```

The source and target virtual machine definitions differ in two key aspects:

* The target's `runStrategy` is set to `WaitAsReceiver`.
* The target virtual machine includes an annotation specifying the `runStrategy` after migration, which should be the same as the source, in this case it is `Always`.

> Any additional resources beyond the virtual machine disk, such as `Secret`s or `Configmap`s, must also be created on the target cluster. Upon creation of the `VirtualMachine` resource, KubeVirt will generate a corresponding receiving `VirtualMachineInstace`, which can then be referenced in the target migration resource.

Reviewer: maybe this is a good chance to highlight that an orchestration layer outside kubevirt to ensure these resources exist in the source cluster, and that is a future goal (is it?..)

Is there any ongoing work about this orchestration layer ? The network team has some requirements for it :)

1. **Accessing the KubeVirt CR on the Target Cluster:** If you have access to the KubeVirt CR, a new field named `synchronizationAddresses` will provide a list of address/port combinations for the `connectURL`.  
2. **After Target Migration Resource Creation:** If you lack KubeVirt CR access, once the target migration resource is created, its status will contain a `synchronizationAddresses` field with the same list of address/port combinations.

The key distinction between these options is that Method 1 allows you to obtain the connection URL in advance, enabling the creation of the source migration resource simultaneously with or prior to the target migration resource.

Once the connection URL is acquired, its value should be placed in the `connectURL` field of the source migration resource.

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstanceMigration
metadata:
  name: example-source-migration
spec:
  sendTo:
	connectURL: "<synchronization address>"
	migrationID: <unique identifier, same as target migration resource migrationID>
  vmiName: <name of the source vmi>
```

### Configuring KubeVirt CA with cross cluster live migration

> When migrating between two clusters, each cluster will have its own installation of KubeVirt. Each KubeVirt installation will create and use its own Certificate Authority (CA). This CA is used to secure communication between the synchronization controllers and the virt-handlers during migration. For each cluster to trust the other cluster, the public CA keys must be exchanged between the clusters.

Author: Sorry no condition to check, you have to check the `kubevirt-ca` configMap to ensure your CA is in there.

The public CA key is stored in a `ConfigMap` in the kubevirt namespace. This `ConfigMap` is named `kubevirt-ca`. The following is an example of this `ConfigMap`:

```yaml
apiVersion: v1
data:
  ca-bundle: |
    -----BEGIN CERTIFICATE-----
...
    -----END CERTIFICATE-----
    -----BEGIN CERTIFICATE-----
...
    -----END CERTIFICATE-----
kind: ConfigMap
metadata:
  name: kubevirt-ca
  namespace: kubevirt
```

> In the same namespace, there will be another `ConfigMap` named `kubevirt-external-ca` which will have a `ca-bundle` field as part of its data. In order for cluster1 to trust the CA from cluster2, take the ca-bundle from cluster2's `kubevirt-ca` `ConfigMap` and copy the `ca-bundle` into cluster1's `kubevirt-external-ca` and vice versa. After you update the `kubevirt-external-ca`, it will automatically clear out the contents of the `kubevirt-external-ca` but copy the contents into the `kubevirt-ca` `ConfigMap`. It will remove invalid, duplicate and expired keys before doing so.

Reviewer 2: ```suggestion
In the same namespace, there will be another `ConfigMap` named `kubevirt-external-ca` which will have a `ca-bundle` field as part of its data. In order for Cluster 1 to trust the CA from Cluster 2, take the ca-bundle from Cluster 2's `kubevirt-ca` `ConfigMap` and copy the `ca-bundle` into Cluster 1's `kubevirt-external-ca` and vice versa. After you update the `kubevirt-external-ca`, it will automatically clear out the contents of the `kubevirt-external-ca` but copy the contents into the `kubevirt-ca` `ConfigMap`. It will remove invalid, duplicate and expired keys before doing so.
```
To match diagram and be consistent with earlier name

# Cancelling decentralized live migration

To cancel an active decentralized live migration, delete the corresponding migration resources from both the source and target cluster/namespace. This action signals KubeVirt to cancel the migration. The cancellation process for decentralized live migrations operates under the same principles as standard migration cancellations. There may be a short delay during which the migration could potentially finish, at which point cancelling has no affect.

# Migration failure

Unlike standard live migrations that automatically retry until completion, failed decentralized live migrations do not. When a decentralized live migration fails, the source virtual machine continues to run on the source, and the target virtual machine enters a failed state. To reattempt the migration, you must manually delete both the source and target `VirtualMachineInstanceMigration` resources, as well as the target virtual machine.

# Limitations

Since decentralized live migration is a variation of storage live migration it has the same limitations:

> * Shareable disks are not supported, as we cannot guarantee data consistency with multiple writers.  

Reviewer 3: ```suggestion
* Shareable disks are not supported, including the VM state PVC (for persistent TPM/EFI), as we cannot guarantee data consistency with multiple writers.  
```
